### PR TITLE
[Enhancement] derive predicate to child based on join on predicate

### PR DESCRIPTION
**Pull request by packy92**
## Why I'm doing:
support move  predicate in one child to the other child based on the join on predicate. A simple case is:
```
select * from (select * from t1 where colA = 1) join t2 on t1.colA = t2.colA
``` 
will be optimized like
```
select * from (select * from t1 where colA = 1) join (select * from t2 where colA = 1) on t1.colA = t2.colA
```

## What I'm doing:
1. use `ValuePropertyDeriver` to dervie value property info implied in the predicate. For example, the predicate `(colA = 1 and colB = 2 or colA > 5)` implies colA must in `colA = 1 or colA > 5` while colB can any value.
2. use `RangeExtractor` to extrat the range info from pedicate. For example, the range info for `(colA = 1 and colB = 2 or colA > 5)` is [1, +∞)
3. use `OnPredicateMoveAroundRule` to move around predicate based on children valueProperty info and on predicate.

### Benckmark
SSB 1G dataset, 1 * 4C8G FE, 3 * 4C*8G BE, 50 concurrency
Use a sql with flat table and a complex predicate:
```
with t1 as (select * from lineorder_flat where lo_orderdate = '1992-01-01' and lo_orderkey = 159266 and lo_linenumber = 1 and lo_custkey = 19813 and lo_partkey = 163961 and lo_suppkey = 1540 and lo_orderpriority = '4-NOT SPECI' and lo_shippriority = 0 and lo_quantity = 4 and lo_extendedprice = 809984),
t2 as (select * from lineorder_flat where lo_orderkey between 150000 and 200000)
select * from t1 join lineorder_flat t3 join t2 on t1.lo_orderdate = t2.lo_orderdate and t1.lo_suppkey = t2.lo_suppkey and t1.lo_custkey = t3.lo_custkey;
```
When `set global enable_predicate_move_around = false;`
The explain qps is 307.7. FE cpu is almost 100%
The query qps is 7.1.

When `set global enable_predicate_move_around = true;`
The explain qps is 303.8. FE cpu is almost 100%
The query qps is 8.4.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/Operator.java [hunk 1]
```diff
@@ -181,10 +185,26 @@ public RowOutputInfo getRowOutputInfo(List<OptExpression> inputs) {
         return rowOutputInfo;
     }
 
+    public ValueProperty getValueProperty(List<OptExpression> inputs) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/Operator.java:188]: why not implement the interface by visit mode?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            }
+        }
+        return new ValueProperty(newValueMap);
+    }
+
+    public static class ValueWrapper {
+        private ScalarOperator predicateDesc;
+
+        @Nullable
+        private RangeExtractor.RangeDescriptor valueDesc;
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java:101]: only support range?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CompoundPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+public class ValueProperty {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java:28]: add comment

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValuePropertyDeriver.java [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+
+    @Override
+    public Map<ScalarOperator, ScalarOperator> visit(ScalarOperator scalarOperator, Void context) {
+        return Maps.newHashMap();
+    }
+
+    @Override
+    public Map<ScalarOperator, ScalarOperator> visitBinaryPredicate(BinaryPredicateOperator binaryPredicate, Void context) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValuePropertyDeriver.java:50]: the implement like RangeExtractor?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+import com.starrocks.sql.optimizer.rule.RuleType;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class OnPredicateMoveAroundRule extends TransformationRule {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java:45]: pushDownPredicate also generates predicates, would this lead predicate to duplication?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/Operator.java (hunk 1 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/Operator.java:188]: I think we may need get this property in other rules.  For example, we may want replcae column to constant

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java (hunk 2 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java:101]: Yes. We mainly use the range info to derive `<=, >=` predicate based on `<, >, <=, >=` on predicate. If it's a `=` on predicate, we use the predicateDesc to dervie new preidcate.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValuePropertyDeriver.java (hunk 4 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValuePropertyDeriver.java:50]: A little difference.  We primarily use this to generate new predicates to describe this column. For example, `(colA > 1 and coB = 1) or (colA = -1 and colB = 2)` will generate two predicates, `colA > 1 or colA = -1` and `colB = 1 or colB = 2`, respectively, to describe the possible values of colA and colB. Then, we yeild range information from the new predicates.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java (hunk 5 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java:45]: we have `removeRedundantPredicate` method.  `testRedundantPredicate` shows some cases.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 6]
```diff
@@ -509,6 +511,7 @@ private OptExpression logicalRuleRewrite(
             deriveLogicalProperty(tree);
         }
 
+        ruleRewriteDownTop(tree, rootTaskContext, OnPredicateMoveAroundRule.INSTANCE);
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:514]: Why needs to dow-top iterate? just avoid getting `deriveValueProperty ` multi times, any other reasons?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+    private Map<ScalarOperator, ValueWrapper> valueMap;
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java:30]: What't the meaning of `valueMap`? A ScalarOperator and its associated RangeDescriptor?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValuePropertyDeriver.java (hunk 4 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValuePropertyDeriver.java:50]: emm, I think RangeExtractor could extract `colA >= -1` and `colB in (1, 2)`  from `(colA > 1 and coB = 1) or (colA = -1 and colB = 2)` directly ?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+        super(type, pattern);
+    }
+
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        if (!context.getSessionVariable().isEnablePredicateMoveAround()) {
+            return false;
+        }
+        LogicalJoinOperator joinOperator = input.getOp().cast();
+        if (joinOperator.getJoinType().isFullOuterJoin() || joinOperator.getJoinType().isCrossJoin()) {
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java:61]: What if semi/anti join?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValuePropertyDeriver.java (hunk 4 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValuePropertyDeriver.java:50]:
```
 private static class RangeExtractor extends ScalarOperatorVisitor<Void, Void> {
        private final Map<ScalarOperator, ValueDescriptor> descMap = Maps.newHashMap();
```
The RangeExtractor just extract the valueDescriptor.  We need obtain the new predicate like `colA > 1 or colA = -1` which will help dervie predicate based on join on predicate like `colA = colB`.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java (hunk 7 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/property/ValueProperty.java:30]: yes. key is the scalarOperator, value is its valueinfo including a predicate associated with it and its rangeinfo.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java (hunk 6 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:514]: select * from (select * from t1 where v1 = 1) join (select * from t2) on t1.v1 = t2.v1 join (select * from t3) on t2.v1 = t3.v1.  If we use the topdown way, we need use the rule iteratively. Use down top way, we just use the rule once. I think this maybe more efficient.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java (hunk 8 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java:61]: Let me add it.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+                result = OptExpression.create(joinOperator,
+                        Lists.newArrayList(OptExpression.create(filter, input.inputAt(0)), input.inputAt(1))
+                );
+            } else {
+                LogicalFilterOperator toLeftFilter = new LogicalFilterOperator(toLeftPredicate);
+                LogicalFilterOperator toRightFilter = new LogicalFilterOperator(toRightPredicate);
+                result = OptExpression.create(joinOperator,
+                        Lists.newArrayList(OptExpression.create(toLeftFilter, input.inputAt(0)),
+                                OptExpression.create(toRightFilter, input.inputAt(1)))
+                );
```
[the remark this answers is not in the document]
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java:116]: this is to push filter to both children.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java (hunk 8 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java:61]: semi join is added. anti join the derived predicate may helps little, so I skip it.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+        }
+        if (rewriteResult == null) {
+            return null;
+        }
+
+        rewriteResult.setIsPushdown(true);
+        return removeRedundantPredicate(offspring, rewriteResult, existValueProperty);
+    }
+
+    private List<ScalarOperator> distinctPredicates(List<ScalarOperator> predicates) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java:204]: private List<ScalarOperator> distinctPredicates(List<ScalarOperator> predicates) {
    return new ArrayList<>(new LinkedHashSet<>(predicates));
}

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+        if (!existValueProperty.contains(offspring)) {
+            return rewriteResult;
+        }
+
+        ValuePropertyDeriver deriver = new ValuePropertyDeriver();
+        ValueProperty newValueProperty = deriver.derive(rewriteResult);
+
+        Range<ConstantOperator> existRange = existValueProperty.getValueWrapper(offspring).getValueDesc().getRange();
+        Range<ConstantOperator> newRange = newValueProperty.getValueWrapper(offspring).getValueDesc().getRange();
+        if (existRange == null) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java:229]: if (newRange == null || (existRange != null && !newRange.encloses(existRange))) {
        return null;
    }

    return rewriteResult;

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteDownTopTask.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+public class RewriteDownTopTask extends RewriteTreeTask {
+
+    public RewriteDownTopTask(TaskContext context, OptExpression root,
+                              List<Rule> rules) {
+        super(context, root, rules, true);
+    }
+
+    @Override
+    protected void rewrite(OptExpression parent, int childIndex, OptExpression root) {
+        // prune cte column depend on prune right child first
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteDownTopTask.java:39]: why right child first?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteDownTopTask.java [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            OptimizerTraceUtil.logApplyRule(context.getOptimizerContext(), rule, root, result);
+
+            if (result.isEmpty()) {
+                continue;
+            }
+
+            parent.getInputs().set(childIndex, result.get(0));
+            root = result.get(0);
+            change++;
+            deriveLogicalProperty(root);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteDownTopTask.java:67]: Could directly call super.rewrite(OptExpression parent, int childIndex, OptExpression root)?

File: fe/fe-core/src/test/java/com/starrocks/planner/TablePruningCTETest.java [hunk 14]
```diff
@@ -161,6 +161,7 @@ public void testUpdateAggregationPreventPruning() {
     public void testUpdateWithPredicates() {
         String q = getSqlList("sql/tpch_pk_tables/", "q5").get(0);
         String plan = checkHashJoinCountWithOnlyRBO(q, 3);
+        System.out.println(plan);
```
Comment (kangkaisen) [inline fe/fe-core/src/test/java/com/starrocks/planner/TablePruningCTETest.java:164]: remove it

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+            return rewriteResult;
+        }
+    }
+
+    // join on predicate left_tbl_col < right_tbl_col
+    // if we want to derive predicate to left, we need obtain the upper bound value of right_tbl_col
+    // if we want to derive predicate to right, we need obtain the lower bound value of left_tbl_col
+    private ScalarOperator deriveLessPredicate(ScalarOperator offspring,
+                                               Range<ConstantOperator> range, boolean toLeft) {
+        if (toLeft) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/OnPredicateMoveAroundRule.java:245]: if (toLeft && range.hasUpperBound()) {
        return new BinaryPredicateOperator(BinaryType.LE, offspring, range.upperEndpoint());
    } else if (!toLeft && range.hasLowerBound()) {
        return new BinaryPredicateOperator(BinaryType.GE, offspring, range.lowerEndpoint());
    }

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteDownTopTask.java (hunk 13 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/task/RewriteDownTopTask.java:67]: We need rewrite the children first. So cannot use the parent code.

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by chenminghua8** [comment]
The test in "fe/fe-core/src/test/java/com/starrocks/planner/OnPredicateMoveAroundRuleTest.java" cannot pass, why merge it?

**Pull request merged**